internal/util/basesubjectset.go

```diff
@@ -224,32 +243,611 @@ func (bss BaseSubjectSet[T]) UnionWithSet(other BaseSubjectSet[T]) {
+	return wst.wildcard != nil
+}
+
+func (wst *wildcardSubjectTracker[T]) asSlice() []T {
+	if wst.wildcard == nil {
+		return []T{}
+	}
+
+	return []T{*wst.wildcard}
+}
+
+func (wst *wildcardSubjectTracker[T]) clone() *wildcardSubjectTracker[T] {
+	return &wildcardSubjectTracker[T]{
+		constructor: wst.constructor,
+		wildcard:    wst.wildcard,
+	}
+}
+
+func (wst *wildcardSubjectTracker[T]) unionWildcard(foundWildcard T) {
+	// If there is no existing wildcard, use the one that was found.
+	if wst.wildcard == nil {
+		wst.wildcard = &foundWildcard
+		return
+	}
+
+	// Otherwise, union together the conditionals for the wildcards and *intersect* their exclusion
+	// sets. Exclusion sets are intersected because if an exclusion is missing from one wildcard
+	// but not the other, the missing element will be, by definition, in that other wildcard.
+	//
+	// Examples:
+	//  {*} + {*} => {*}
+	//  {* - {user:tom}} + {*} => {*}
+	//  {* - {user:tom}} + {* - {user:sarah}} => {*}
+	//  {* - {user:tom, user:sarah}} + {* - {user:sarah}} => {* - {user:sarah}}
+	//  {*}[c1] + {*} => {*}
+	//  {*}[c1] + {*}[c2] => {*}[c1 || c2]
+	existing := *wst.wildcard
+
+	exisingConcreteExclusions := newConcreteSubjectSet[T](wst.constructor)
+	for _, excludedSubject := range existing.GetExcludedSubjects() {
+		exisingConcreteExclusions.unionConcrete(excludedSubject)
+	}
+
+	foundConcreteExclusions := newConcreteSubjectSet[T](wst.constructor)
+	for _, excludedSubject := range foundWildcard.GetExcludedSubjects() {
+		foundConcreteExclusions.unionConcrete(excludedSubject)
+	}
+
+	exisingConcreteExclusions.intersectDiffWithConcrete(foundConcreteExclusions)
+
+	constructed := wst.constructor(
+		tuple.PublicWildcard,
+		shortcircuitedOr(existing.GetCaveatExpression(), foundWildcard.GetCaveatExpression()),
+		exisingConcreteExclusions.asSlice(),
+		existing,
+		foundWildcard)
+	wst.wildcard = &constructed
+}
+
+func (wst *wildcardSubjectTracker[T]) unionConcrete(concrete T) {
+	if wst.wildcard == nil {
+		// Nothing to do if no wildcard present.
+		return
+	}
+
+	// If the concrete is in the exclusion set, remove it if not conditional. Otherwise, mark
+	// it as conditional.
+	//
+	// Examples:
+	//  {*} | {user:tom} => {*} (and user:tom in the concrete)
+	//  {* - {user:tom}} | {user:tom} => {*} (and user:tom in the concrete)
+	//  {* - {user:tom}[c1]} | {user:tom}[c2] => {* - {user:tom}[c1 && !c2]} (and user:tom in the concrete)
+	existing := *wst.wildcard
+	updatedExclusions := make([]T, 0, len(existing.GetExcludedSubjects()))
+	for _, existingExclusion := range existing.GetExcludedSubjects() {
+		if existingExclusion.GetSubjectId() == concrete.GetSubjectId() {
+			// If the conditional on the concrete is empty, then the concrete is always present, so
+			// we remove the exclusion entirely.
+			if concrete.GetCaveatExpression() == nil {
+				continue
+			}
+
+			// Otherwise, the conditional expression for the new exclusion is the existing expression ||
```

Comment:
code coverage shows this is not being tested. This triggers it, but fails

```go
	{
			"add caveated sub to wildcard with caveated exclusion",
			[]*v1.FoundSubject{
				cwc(nil, csub("foo", caveatexpr("testcaveat"))),
			},
			csub("foo", caveatexpr("testcaveat")),
			true,
			[]*v1.FoundSubject{
				cwc(nil, csub("foo", caveatexpr("testcaveat"))),
			},
		},
```

Reply:
I added this test case (and a few others), but had to adjust the expected results to match